### PR TITLE
added possibility to delete artifact with specified name pattern

### PR DESCRIPTION
**Pull request by wotd**

### CODE REVIEW

File: cleanup/artifactCleanup/artifactCleanup.groovy [hunk 1]
```diff
@@ -78,27 +85,29 @@ executions {
...
-log.info "Schedule job policy list: $config.policies"
-
 config.policies.each{ policySettings ->
     def cron = policySettings[ 0 ] ? policySettings[ 0 ] as String : ["0 0 5 ? * 1"]
     def repos = policySettings[ 1 ] ? policySettings[ 1 ] as String[] : ["__none__"]
     def months = policySettings[ 2 ] ? policySettings[ 2 ] as int : 6
     def paceTimeMS = policySettings[ 3 ] ? policySettings[ 3 ] as int : 0
     def dryRun = policySettings[ 4 ] ? policySettings[ 4 ] as Boolean : false
     def disablePropertiesSupport = policySettings[ 5 ] ? policySettings[ 5 ] as Boolean : false
+    def keepRelease = policySettings[ 6 ] ? policySettings[ 6 ] as Boolean : false
```
Comment (potasiak) [inline cleanup/artifactCleanup/artifactCleanup.groovy:95]: Is `as Boolean` type cast working correctly here?

**Review by potasiak — APPROVED** [review]

File: cleanup/artifactCleanup/artifactCleanup.groovy (hunk 1 above)
Comment (wotd) [reply cleanup/artifactCleanup/artifactCleanup.groovy:95]: surprisingly, yes :)

File: cleanup/artifactCleanup/artifactCleanup.groovy [hunk 2]
```diff
@@ -39,16 +42,20 @@ class Global {
 // curl -i -uadmin:password -X POST "http://localhost:8081/artifactory/api/plugins/execute/cleanupCtl?params=command=stop"
 // curl -i -uadmin:password -X POST "http://localhost:8081/artifactory/api/plugins/execute/cleanupCtl?params=command=adjustPaceTimeMS;value=-1000"
 
+
 def pluginGroup = 'cleaners'
+def config = new ConfigSlurper().parse(new File(ctx.artifactoryHome.haAwareEtcDir, PROPERTIES_FILE_PATH).toURL())
```
Comment (martinm82) [inline cleanup/artifactCleanup/artifactCleanup.groovy:47]: Hm, could you please check whether this option is available as well in non-ha setups?

File: cleanup/artifactCleanup/artifactCleanup.groovy [hunk 3]
```diff
@@ -39,16 +42,20 @@ class Global {
...
 
 executions {
     cleanup(groups: [pluginGroup]) { params ->
         def months = params['months'] ? params['months'][0] as int : 6
         def repos = params['repos'] as String[]
-        def dryRun = params['dryRun'] ? params['dryRun'][0] as boolean : false
-        def disablePropertiesSupport = params['disablePropertiesSupport'] ? params['disablePropertiesSupport'][0] as boolean : false
+        def dryRun = params['dryRun'] ? params['dryRun'][0].toBoolean() : false
```
Comment (martinm82) [inline cleanup/artifactCleanup/artifactCleanup.groovy:53]: Why was this change needed? Does look weird mixing styles (`as <TYPE>` vs. `.toType()`)

File: cleanup/artifactCleanup/artifactCleanup.groovy [hunk 4]
```diff
@@ -78,27 +85,29 @@ executions {
...
-
 config.policies.each{ policySettings ->
     def cron = policySettings[ 0 ] ? policySettings[ 0 ] as String : ["0 0 5 ? * 1"]
     def repos = policySettings[ 1 ] ? policySettings[ 1 ] as String[] : ["__none__"]
     def months = policySettings[ 2 ] ? policySettings[ 2 ] as int : 6
     def paceTimeMS = policySettings[ 3 ] ? policySettings[ 3 ] as int : 0
     def dryRun = policySettings[ 4 ] ? policySettings[ 4 ] as Boolean : false
     def disablePropertiesSupport = policySettings[ 5 ] ? policySettings[ 5 ] as Boolean : false
+    def keepRelease = policySettings[ 6 ] ? policySettings[ 6 ] as Boolean : false
+    def releaseRegex = policySettings[ 7 ] ? policySettings[ 7 ] as Pattern : ~/.*-\d+\.\d+\.\d+\.*/
```
Comment (martinm82) [inline cleanup/artifactCleanup/artifactCleanup.groovy:96]: Store the default pattern in a variable and re-use it in all other places

File: cleanup/artifactCleanup/artifactCleanup.groovy [hunk 5]
```diff
@@ -78,27 +85,29 @@ executions {
...
         }
     }
 }
 
-private def artifactCleanup(int months, String[] repos, log, paceTimeMS, dryRun = false, disablePropertiesSupport = false) {
-    log.info "Starting artifact cleanup for repositories $repos, until $months months ago with pacing interval $paceTimeMS ms, dryrun: $dryRun, disablePropertiesSupport: $disablePropertiesSupport"
+private def artifactCleanup(int months, String[] repos, log, paceTimeMS, dryRun = false, disablePropertiesSupport = false, keepRelease = false, releaseRegex = ~/.*-\d+\.\d+\.\d+\.*/) {
+    log.info "Starting artifact cleanup for repositories $repos, until $months months ago with pacing interval $paceTimeMS ms, dryrun: $dryRun, disablePropertiesSupport: $disablePropertiesSupport, keepRelease: $keepRelease, releaseRegex: $releaseRegex"
 
     // Create Map(repo, paths) of skiped paths (or others properties supported in future ...)
```
Comment (martinm82) [inline cleanup/artifactCleanup/artifactCleanup.groovy:112]: Unrelated change: 

Typo: skiped vs. skipped
Typo: others vs. other

File: cleanup/artifactCleanup/artifactCleanup.groovy [hunk 6]
```diff
@@ -141,17 +150,25 @@ private def artifactCleanup(int months, String[] repos, log, paceTimeMS, dryRun
                 cntNoDeletePermissions++
             }
             if (dryRun) {
-                log.info "Found $it, $cntFoundArtifacts/$artifactsCleanedUp.size total $bytesFound bytes"
-                log.info "\t==> currentUser: ${security.currentUser().getUsername()}"
-                log.info "\t==> canDelete: ${security.canDelete(it)}"
-
+                if (checkName(keepRelease, releaseRegex, it)) {
+                    log.info "Found $it, $cntFoundArtifacts/$artifactsCleanedUp.size total $bytesFound bytes"
```
Comment (martinm82) [inline cleanup/artifactCleanup/artifactCleanup.groovy:154]: The 'Found', 'currentUser' and 'canDelete' log calls can be moved out of the if/else block (DRY)

File: cleanup/artifactCleanup/artifactCleanup.properties [hunk 7]
```diff
@@ -1,6 +1,6 @@
 // Support different policies for different repos. Parameters between '|' are optional
-// Policy syntax: [ <cron time>, [ <repo list> ], months |, <pause time>, <true|false dryrun> |, <true|false disablePropertiesSupport> | }
-// Example: [ "0 0 5 ? * 1", [ "libs-releases-local" ], 3, 500, true, true ],
+// Policy syntax: [ <cron time>, [ <repo list> ], months |, <pause time>, <true|false dryrun> |, <true|false disablePropertiesSupport> |, <true|false> keepRelease |, releaseRegex | }
```
Comment (martinm82) [inline cleanup/artifactCleanup/artifactCleanup.properties:2]: Should be
```
<true|false keepRelease> |, [...]
```

**Review by martinm82 — COMMENTED** [review]
Needs some work

File: cleanup/artifactCleanup/artifactCleanup.groovy (hunk 2 above)
Comment (wotd) [reply cleanup/artifactCleanup/artifactCleanup.groovy:47]: I am testing it on artifactory deployed as single instance. it is definitely non-ha setup :)

File: cleanup/artifactCleanup/artifactCleanup.groovy (hunk 3 above)
Comment (wotd) [reply cleanup/artifactCleanup/artifactCleanup.groovy:53]: Because for these particular params it was not casted properly. For string, as boolean return true if string is not empty. It was changed because if you put for example dryRun=false in curl it will be true anyway.

File: cleanup/artifactCleanup/artifactCleanup.groovy (hunk 6 above)
Comment (wotd) [reply cleanup/artifactCleanup/artifactCleanup.groovy:154]: Yes, I am doing refactoring of this part

File: cleanup/artifactCleanup/artifactCleanup.groovy (hunk 2 above)
Comment (martinm82) [reply cleanup/artifactCleanup/artifactCleanup.groovy:47]: thx

File: cleanup/artifactCleanup/artifactCleanup.groovy (hunk 3 above)
Comment (martinm82) [reply cleanup/artifactCleanup/artifactCleanup.groovy:53]: Thanks a lot. It would make sense to push the toBoolean change to the upstream project as a separate PR since this is definitely an issue

File: cleanup/artifactCleanup/artifactCleanup.groovy [hunk 8]
```diff
@@ -78,27 +85,29 @@ executions {
...
-            artifactCleanup( months, repos, log, paceTimeMS, dryRun, disablePropertiesSupport )
+            log.info "Policy settings for scheduled run at($cron): repo list($repos), months($months), paceTimeMS($paceTimeMS) dryrun($dryRun) disablePropertiesSupport($disablePropertiesSupport) keepRelease($keepRelease), releaseRegex($releaseRegex)"
+            artifactCleanup( months, repos, log, paceTimeMS, dryRun, disablePropertiesSupport, keepRelease, releaseRegex )
         }
     }
 }
 
-private def artifactCleanup(int months, String[] repos, log, paceTimeMS, dryRun = false, disablePropertiesSupport = false) {
-    log.info "Starting artifact cleanup for repositories $repos, until $months months ago with pacing interval $paceTimeMS ms, dryrun: $dryRun, disablePropertiesSupport: $disablePropertiesSupport"
+private def artifactCleanup(int months, String[] repos, log, paceTimeMS, dryRun = false, disablePropertiesSupport = false, keepRelease = false, releaseRegex = ~/.*-\d+\.\d+\.\d+\.*/) {
```
Comment (martinm82) [inline cleanup/artifactCleanup/artifactCleanup.groovy:109]: Why is the default value for keepRelease set to false? Wouldn't it be safer to set it to true to avoid accidental removal?

File: cleanup/artifactCleanup/artifactCleanup.groovy [hunk 9]
```diff
@@ -218,3 +235,7 @@ private def getSkippedPaths(String[] repos) {
     log.info "Elapsed time to retrieve paths to skip: " + duration
     return skip
 }
+
+private def checkName(keepRelease, releaseRegex, artifactName) {
```
Comment (martinm82) [inline cleanup/artifactCleanup/artifactCleanup.groovy:239]: I think the function name is as well not good enough. The function returns a boolean so the name should somehow indicate that (similar to canDelete, isAvailable, etc.)

Comment (martinm82) [inline cleanup/artifactCleanup/artifactCleanup.groovy:239]: So if I set `keepRelease` to `true` than the regex check is still evaluated and my release artifact might still get deleted
